internal/testutils/pipe_listener.go

```diff
@@ -0,0 +1,80 @@
+/*
+ *
+ * Copyright 2018 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package testutils
+
+import (
+	"errors"
+	"net"
+	"time"
+)
+
+type pipeAddr struct{}
+
+func (p pipeAddr) Network() string { return "pipe" }
+func (p pipeAddr) String() string  { return "pipe" }
+
+// pipeListener is a listener with an unbuffered pipe. Each write will complete only once the other side reads. It
+// should only be created using NewPipeListener.
+type pipeListener struct {
+	C chan chan<- net.Conn
+}
+
+// NewPipeListener creates a new pipe listener.
+func NewPipeListener() *pipeListener {
+	return &pipeListener{
+		C: make(chan chan<- net.Conn),
+	}
+}
+
+// Accept accepts a connection.
+func (p *pipeListener) Accept() (net.Conn, error) {
+	connChan, ok := <-p.C
+	if !ok {
+		return nil, errors.New("closed")
+	}
+	c1, c2 := net.Pipe()
+	connChan <- c1
+	close(connChan)
+	return c2, nil
+}
+
+// Close closes the listener.
+func (p *pipeListener) Close() error {
+	// We don't close p.C, because it races with the client reconnect.
```

Comment:
How about: lock, close p.c, unlock.  In Dial: lock, check that p.c is open, make a new chan and push onto p.c, unlock?  I think that avoids any potential races.